// gremlin/lib/WebSocketGremlinConnection.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _events = require('events');

var _ws = require('ws');

var _ws2 = _interopRequireDefault(_ws);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class WebSocketGremlinConnection extends _events.EventEmitter {
  constructor({ port, host, path, ssl, rejectUnauthorized }) {
    super();

    this.open = false;

    const address = `ws${ssl ? 's' : ''}://${host}:${port}${path}`;
    const options = {
      rejectUnauthorized
    };

    this.ws = new _ws2.default(address, null, options);

    this.ws.onopen = () => this.onOpen();
    this.ws.onerror = err => this.handleError(err);
    this.ws.onmessage = message => this.handleMessage(message);
    this.ws.onclose = event => this.onClose(event);
    this.ws.binaryType = 'arraybuffer';
  }

  onOpen() {
    this.open = true;
    this.emit('open');
  }

  handleError(err) {
    this.emit('error', err);
  }

  handleMessage(message) {
    this.emit('message', message);
  }

  onClose(event) {
    this.open = false;
    this.emit('close', event);
  }

  sendMessage(message) {
    this.ws.send(message, { mask: true, binary: true }, err => {
      if (err) {
        this.handleError(err);
      }
    });
  }
}
exports.default = WebSocketGremlinConnection;
