// zer/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.gremlin = exports.groovy = exports.createChainCreator = undefined;

var _factories = require('./factories');

Object.defineProperty(exports, 'createChainCreator', {
  enumerable: true,
  get: function () {
    return _factories.createChainCreator;
  }
});
exports.inspectChain = inspectChain;
exports.renderChain = renderChain;
exports.innerChain = innerChain;
exports.inspectRenderer = inspectRenderer;
exports.inspectSyntax = inspectSyntax;

var _groovy = require('./lang/groovy');

var _groovy2 = _interopRequireDefault(_groovy);

var _inline = require('./render/inline');

var _inline2 = _interopRequireDefault(_inline);

var _gremlinServer = require('./render/gremlin-server');

var _gremlinServer2 = _interopRequireDefault(_gremlinServer);

var _chainBuilder = require('./chain-builder');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const groovy = exports.groovy = (0, _factories.createChainCreator)(_inline2.default, _groovy2.default);
const gremlin = exports.gremlin = (0, _factories.createChainCreator)(_gremlinServer2.default, _groovy2.default);

function inspectChain(chainProxy) {
  return chainProxy[_chainBuilder.inspectSymbol];
}

function renderChain(chainProxy) {
  return chainProxy[_chainBuilder.renderSymbol];
}

function innerChain(chainProxy) {
  return chainProxy[_chainBuilder.chainSymbol];
}

function inspectRenderer(chain) {
  return chain[_chainBuilder.rendererSymbol];
}

function inspectSyntax(chain) {
  return chain[_chainBuilder.syntaxSymbol];
}

exports.default = {
  groovy,
  gremlin
};
